r/R/transforms.R

```diff
@@ -812,3 +812,84 @@ multiply_cols <- function(mat, vec) {
   assert_true(length(vec) == ncol(mat))
   t(t(mat) * vec)
 }
+
+
+
+#################
+# Linear residual
+#################
+
+setClass("TransformLinearResidual", contains = "TransformedMatrix")
+setMethod("iterate_matrix", "TransformLinearResidual", function(x) {
+  res <- iterate_matrix(x@matrix)
+  
+  if (any(nrow(x@row_params) == 0, nrow(x@col_params) == 0)) {
+    return(res)
+  }
+  iterate_matrix_linear_residual_cpp(res, x@row_params, x@col_params)
+})
+setMethod("short_description", "TransformLinearResidual", function(x) {
+  # Return multiple lines, one for each transform active
+  res <- short_description(x@matrix)
+  if (any(nrow(x@row_params) == 0, nrow(x@col_params) == 0)) {
+    res <- c(res, sprintf("Linear regress out 0 variable"))
+    return(res)
+  }
+  res <- c(res, sprintf("Linear regress out %d variable(s)", nrow(x@row_params) - 1))
```

Comment:
This is a picky comment but I think will make the user experience a lot easier: could we make it so this will print out some of the actual variable names that are getting regressed out? I think this will require:

1. adding a slot to store the variable names
2.  initializing the variable names in `regress_out`
3.  printing it out here via a call to `pretty_print_vector(x@var_names)`

I don't think we can rely on `nrow(x@row_params)` because if someone passes in a factor to regress out, then `model.matrix` will expand it into indicator variables, but logically its just a single variable getting regressed out.

```suggestion
  res <- c(res, sprintf("Linear regress out %d variable(s)%s", length(x@var_names), pretty_print_vector(x@var_names, prefix=": ")))
```